File: composite/composite.py
from abc import ABCMeta, abstractmethod
import inspect


class Component(metaclass=ABCMeta):
    @abstractmethod
    def get_price(self):
        pass

    def print(self, price, depth):
        indent = ' ' * (depth - 2) * 2
        print('{}{}: {}'.format(indent, self.name, price))


class Product(Component):
    def __init__(self, name, price):
        self.name = name
        self.price = price
    
    def get_price(self):
        self.print(self.price, len(inspect.stack()))
        return self.price


class CompositeProduct(Component):
    def __init__(self, name, children=list()):
        self.name = name
        self.children = children

    def get_price(self):
        price = sum(child.get_price() for child in self.children)
        self.print(price, len(inspect.stack()))
        return price


turbocharger = Product('turbocharger', 500)
intercooler = Product('intercooler', 200)
engine_block = Product('engine_block', 1000)
engine = CompositeProduct('engine', [turbocharger, intercooler, engine_block])

tire = Product('tire', 50)
rim = Product('rim', 100)
wheel = CompositeProduct('wheel', [tire, rim])

gearbox = Product('gearbox', 500)

body = Product('body', 1500)

car = CompositeProduct('car', [body, engine, gearbox, wheel, wheel, wheel, wheel])

car.get_price()

File: composite/composite.js
const Component = {
    getPrice: function() {},
    print: function(price) {console.log(this.name + ': ' + price)},
};

function Product(name, price) {
    this.name = name;
    this.price = price;
    this.getPrice = function() {
        this.print(this.price);
        return this.price;
    };
};
Product.prototype = Component;

function CompositeProduct(name, children) {
    this.name = name;
    this.children = children;
    this.getPrice = function() {
        let price = 0;
        for (let c of children) {
            price += c.getPrice();
        };
        this.print(price);
        return price;
    };
};
CompositeProduct.prototype = Component;

let turbocharger = new Product('turbocharger', 500);
let intercooler = new Product('intercooler', 200);
let engine_block = new Product('engine_block', 1000);
let engine = new CompositeProduct('engine', [turbocharger, intercooler, engine_block]);

let tire = new Product('tire', 50);
let rim = new Product('rim', 100);
let wheel = new CompositeProduct('wheel', [tire, rim]);

let gearbox = new Product('gearbox', 500);

let body = new Product('body', 1500);

let car = new CompositeProduct('car', [body, engine, gearbox, wheel, wheel, wheel, wheel]);

car.getPrice();
